feat: emberplus api expansion

phillipivan opened the pull request:
Expands emberplus api by exposing internal variables, custom variables, and action recorder controls.
Internal variables are exposed as read only string, bool or int parameters, custom variables are coerced to string type parameters as the likely best compromise between the dynamic typing of companion variables and the strict typing of ember plus parameters, they are read/write.

- [x] Get list of internal and custom variables from variable controller
- [x] Define ember plus tree
- [x] Handle `setValue` for Custom Variables
- [x] Handle `actionRecorder` controls and state (mapped via serviceApi)
- [x] Monitor for changes in custom variable count and reinit server (as per changes for page count and button matrix size). NB, At present, on start up, the ember plus server is initialised before the variables are defined it seems and the service needs to be toggled before any variables are available in the tree (internal and custom).
- [x] Listen for variable change events, rather than polling for better efficiency and responsiveness
- [x] Decide how to best handle unstable node paths for variables (between companion restarts)
- [x] Update user docs

phillipivan commented:
Each variable is defined in its own node as behaviour is poor when defined in a flat structure:
![image](https://github.com/user-attachments/assets/110b6eee-f9d5-4dbc-ac15-91266eab95fa)
For now (at least) uptime is also mapped to the identity node. I dont feel strongly about this and am happy to remove:
![image](https://github.com/user-attachments/assets/76fd732e-d9c4-4d51-a5e8-a51aa308ac6b)
Action recorder control - the action count variable is duplicated here for convenience. 
![image](https://github.com/user-attachments/assets/2e1bd0d9-f2e3-43fd-bbcf-4de374b434e5)

phillipivan commented:
@Julusian I would welcome any advice you have regarding the last two issues, and any other feedback you have.

Julusian commented on companion/lib/Service/EmberPlus.ts:
```diff
@@ -100,9 +104,68 @@ export class ServiceEmberPlus extends ServiceBase {
 	}
 
+	private startPolling(): void {
+		this.#pollTimer = setInterval(() => {
```
I'm not keen on this polling, I would prefer this worked by being hooked into the existing variables changed flow. that would allow this to be a bit simpler and avoid having to compare each value.

You could probably cheat on the action recorder values, and look up the variables which expose the same values too in order to avoid needing to poll the status of that.

I wonder, does the ember library support lazy values? It feels a bit wasteful to have to provide values for each node, when it wont be doing anything with them until a consumer requests them. The ideal flow would be to handle the server subscribing and unsubscribing to values, so that only the interesting values have to be provided. May be too much work to achieve, and not worth the effort if it requires digging in the library.

phillipivan replied:
I hope I am able to complete remove the polling once I have the `variable_changed` event listening sorted, but putting the polling in allowed me to get one aspect sorted before moving on to the next.

How would feel about leaving the lazy values stuff for now? I agree it would be the best approach but I suspect it would require work on the emberplus library. If a way is found to do this we could always circle back and implement it.

phillipivan replied:
All polling removed and replaced with event listeners, including for action recorder state.

Julusian commented:
> Monitor for changes in custom variable count and reinit server

I would like to see an event based flow for this. I would be tempted by making `VariablesCustomVariable` extend `EventEmitter`, (see elsewhere for how to do the types for this) with an event perhaps `definitionChanged: [id: string, info: CustomVariableDefinition | null]`. This would need emitting anywhere which changes the info of a variable.
This api can then listen for that event (with a debounce) and reinit (if needed, some changes could be `persistValue` or the description changing)

> the ember plus server is initialised before the variables are defined it seems 

perhaps this is generally wrong, and all the apis are being started too early?  
we don't start the http (admin ui) api until after everything has been initialised, perhaps all these apis should be delayed until the same point in time?

>  Listen for variable change events,

ah, I read the code before your comments..  
Yeah, if you search for `variables_changed` you will see the main flow for how we handle changes elsewhere, it should be easy enough to hook into that. Maybe that event should be expanded to include an array of connectionLabels this originated from, which would allow you to check really quickly if it was an interesting update.

phillipivan commented:
I think I have addressed most of that. There is no more polling whatsoever, all updates are from event listeners. I also added one for the action recorder state.

Server restarts are automated and debounced. This also crudely addresses the initialization order issue.

I am surprised to see that some deleted variables (custom, and internal connection variables) are still returned from the `getConnectionVariableDefinitions` method. 

Such as:
![image](https://github.com/user-attachments/assets/157e0fe8-da08-4caa-854a-fbc238d8fe42)

Which has been deleted:
![image](https://github.com/user-attachments/assets/9963d7ff-06cb-4b28-b426-e0ca26452d9b)

Or:
![image](https://github.com/user-attachments/assets/63540adc-173f-4bfa-8712-6c4abc886ccb)
And New:
![image](https://github.com/user-attachments/assets/973aba89-e562-4a97-b9ef-97ace7210c8e)
From a connection list of:
![image](https://github.com/user-attachments/assets/e8aeb3cf-ff14-4ad6-8987-25f8498fe422)


While I am not especially concerned about no longer relevant variables persisting in the tree - it does bring about a subtle problem, that the node numbers of the variables may change after a full companion restart where the old variables are flushed.

However after a full companion restart we get:
![image](https://github.com/user-attachments/assets/dd509c71-d03b-45a0-92ea-ca02091c7788)
![image](https://github.com/user-attachments/assets/a1c4a6c4-a4f4-407b-8ae4-79e9ae3041b4)

Some thought needs to go into how best to handle this - as the potential for unstable ember paths will eventually cause someone heartache - and its not just potential instability of the paths for those specific variables, but all others that are normally defined after them in the list. @Julusian is it deliberate (I assume so) that these variables persist internally even after they been deleted from the user perspective?

phillipivan commented:
Currently my best idea for the unstable node numbering, which wouldn't really solve it, but perhaps limit the scope of impact would be to sort the internal variables into a few categories connections, pages, surfaces, date, time, system, based on the variable name. 

It will clutter up the code some, and wouldn't actually solve the problem, simply mitigate it somewhat, so I am not sure if it would be worth the effort. Thoughts?

It may be we have to advise to use textual node paths whenever possible.

phillipivan commented on docs/structure.json:
```diff
@@ -260,6 +260,10 @@
 				"label": "RossTalk Control",
 				"file": "5_remote_control/rosstalk_control.md"
 			},
+			{
```
As far as I can tell this is where the new `emberplus_control.md` document needs to be referenced, but I cant get this to appear when building Companion. Is there somewhere else it needs to be referenced, or a different build method required?

phillipivan commented:
I have added some documentation. What do I need to do to have it appear in the `Settings - Protocols` right hand tab?

<img width="925" alt="image" src="https://github.com/user-attachments/assets/3c64ed62-f150-4aef-8729-cce046f937c6" />

Julusian reviewed (approved):
just a couple of minor things

phillipivan commented on companion/lib/Service/ServiceApi.ts:
```diff
@@ -24,7 +29,10 @@ import type { GraphicsController } from '../Graphics/Controller.js'
  */
-export class ServiceApi {
+
+type ServiceApiEvents = VariablesValuesEvents | ActionRecorderEvents | VariablesCustomVariableEvents
```
I think using your suggesting of `Pick<T,K>` means that inheriting any future args from your prior comment will work cleanly? That's the theory I'm going with anyway.

phillipivan commented:
@Julusian did you clock my question about the types for the `t-bar` and `jog` interval variables? Do I need to make these float, or will they work as ints?

phillipivan commented on companion/lib/Service/EmberPlus.ts:
```diff
@@ -104,6 +108,85 @@ export class ServiceEmberPlus extends ServiceBase {
+		this.#serviceApi.on('action_recorder_is_running', (is_running) => {
+			if (this.#server) {
+				//check action recorder status
+				const node = this.#server.getElementByPath('0.4.0')
```
@Julusian I have pulled some of the hard coded paths out to consts, and added a utility function to build the variable paths, similar to the ones for Location and Button. Let me know if you are happy with  this approach.

Julusian commented:
> did you clock my question about the types for the t-bar and jog interval variables? Do I need to make these float, or will they work as ints?

I missed that, I think int should be fine. Tbh maybe this release I should try and finally remove these, as there is a better multi-surface flow that can be used instead now..

> is it deliberate (I assume so) that these variables persist internally even after they been deleted from the user perspective?

No, this isn't deliberate. Its likely just a case of bad handling on our part, forgetting to consider that they might need to be have the values/definitions unset

> Is there a practical way to pull the description of the internal variables? 

Doesnt look like it. Perhaps some stuff needs to be exposed from VariablesInstanceDefinitions to allow this
